clippy_lints/src/misc.rs

```diff
@@ -621,17 +621,25 @@ fn non_macro_local(cx: &LateContext<'_, '_>, res: def::Res) -> bool {
 
 fn check_cast(cx: &LateContext<'_, '_>, span: Span, e: &Expr, ty: &Ty) {
     if_chain! {
-        if let TyKind::Ptr(MutTy { mutbl, .. }) = ty.kind;
+        if let TyKind::Ptr(ref mut_ty) = ty.kind;
         if let ExprKind::Lit(ref lit) = e.kind;
-        if let LitKind::Int(value, ..) = lit.node;
-        if value == 0;
+        if let LitKind::Int(0, _) = lit.node;
         if !in_constant(cx, e.hir_id);
         then {
-            let msg = match mutbl {
-                Mutability::MutMutable => "`0 as *mut _` detected. Consider using `ptr::null_mut()`",
-                Mutability::MutImmutable => "`0 as *const _` detected. Consider using `ptr::null()`",
+            let (msg, sugg_fn) = match mut_ty.mutbl {
+                Mutability::MutMutable => ("`0 as *mut _` detected", "std::ptr::null_mut"),
+                Mutability::MutImmutable => ("`0 as *const _` detected", "std::ptr::null"),
             };
-            span_lint(cx, ZERO_PTR, span, msg);
+            if let Some(ty_snip) = snippet_opt(cx, ty.span) {
+                let sugg = match mut_ty.ty.kind {
+                    TyKind::Infer => String::from(sugg_fn) + "()",
+                    _ => format!("{}::<{}>()", sugg_fn, ty_snip),
+                };
+                span_lint_and_sugg(cx, ZERO_PTR, span, msg, "try", sugg, Applicability::MachineApplicable)
+            } else {
+                // `MaybeIncorrect` as type inference may not work with the suggested code
+                span_lint_and_sugg(cx, ZERO_PTR, span, msg, "try", String::from(sugg_fn) + "()", Applicability::MaybeIncorrect)
```

Comment:
I'd prefer `format!` here
```suggestion
                span_lint_and_sugg(cx, ZERO_PTR, span, msg, "try", format!("{}()", sugg_fn), Applicability::MaybeIncorrect)
```